### PR TITLE
Fix #4761 - Cleanup temporary directory when doing *Schematron* XMLValidation

### LINKED ISSUE #4761: XML validation doesn't clean up after itself
Issue overview
--------------
I am using OpenStudio 3.5's XSD/Schematron validation in BEopt, where we may end up validating thousands of HPXML files during an optimization of EnergyPlus simulations.

The computer ended up crashing because it ran out of disk space. It looks like every individual OpenStudio validation produces a 1MB temp directory:

image

After running just a few BEopt optimizations, there were over 10k directories taking up 12 GB of space.

Can OpenStudio clean up the temp dir after validation?

@jmarrec

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #4761

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/utilities/xml/Test/XMLValidator_GTest.cpp [hunk 1]
```diff
@@ -135,28 +135,37 @@ TEST_F(XMLValidatorFixture, XMLValidator_HPXMLvalidator_Schematron) {
...
+      ASSERT_EQ(1, errors.size()) << "Failed at iteration " << i + 1;
+      EXPECT_EQ(LogLevel::Error, errors[0].logLevel());
+      EXPECT_EQ("xsltValidate: Expected EventType to be 'audit' or 'proposed workscope' or 'approved workscope' or "
+                "'construction-period testing/daily test out' or 'job completion testing/final inspection' or "
+                "'quality assurance/monitoring' or 'preconstruction'",
+                errors[0].logMessage());
+    }
   }
+  // #4761 - XMLValidator's dtor should clean up the tmpDir
+  EXPECT_FALSE(openstudio::filesystem::exists(tmpDir)) << "Expected tmpDir to be deleted: " << tmpDir;
```
Comment (jmarrec) [inline src/utilities/xml/Test/XMLValidator_GTest.cpp:168]: Fails before fix, works after

File: src/utilities/xml/XMLValidator.cpp [hunk 2]
```diff
@@ -198,12 +198,12 @@ XMLValidator::XMLValidator(const openstudio::path& schemaPath) : m_schemaPath(op
     logAndStore(Info, "Treating schema as a Schematron, converting to an XSLT StyleSheet.");
     // Let's use a temporary directory for this, so we avoid having two instances trying to write to the same file and we avoid issues where the
     // directory is write protected.
-    const auto tmpDir = openstudio::filesystem::create_temporary_directory("xmlvalidation");
-    if (tmpDir.empty()) {
-      LOG_AND_THROW(fmt::format("Failed to create a temporary directory for extracting the stylesheets need to transform the Schematron '{}'",
+    m_tempDir = openstudio::filesystem::create_temporary_directory("xmlvalidation");
```
Comment (jmarrec) [inline src/utilities/xml/XMLValidator.cpp:201]: store the temp dir

File: src/utilities/xml/XMLValidator.cpp [hunk 3]
```diff
@@ -212,16 +212,16 @@ XMLValidator::XMLValidator(const openstudio::path& schemaPath) : m_schemaPath(op
...
+XMLValidator::~XMLValidator() {
+  if (m_tempDir) {
+    try {
+      const auto count = openstudio::filesystem::remove_all(m_tempDir.get());
+      logAndStore(Debug, fmt::format("Removed temporary directory with {} files", count));
+    } catch (const std::exception& e) {
+      logAndStore(Warn, fmt::format("Error removing temporary directory at {}, Description: {}", toString(m_tempDir.get()), e.what()));
+    }
+  }
+}
```
Comment (jmarrec) [inline src/utilities/xml/XMLValidator.cpp:224]: Delete temp Dir in dtor

File: src/utilities/xml/XMLValidator.hpp [hunk 4]
```diff
@@ -68,6 +68,11 @@ class UTILITIES_API XMLValidator
    * - `*.xml` or `*.sct` => Schematron (convert to XSLT then validate)
    * - `*.xslt` => Schematron that is already transformed to an XSLT stylesheet */
   explicit XMLValidator(const openstudio::path& schemaPath);
+  ~XMLValidator();
+  XMLValidator(const XMLValidator& other) = default;
+  XMLValidator(XMLValidator&& other) noexcept = default;
+  XMLValidator& operator=(const XMLValidator& other) = default;
+  XMLValidator& operator=(XMLValidator&& other) noexcept = default;
```
Comment (jmarrec) [inline src/utilities/xml/XMLValidator.hpp:75]: Rule of 5

**Review by shorowit — APPROVED** [review]
Haven't reviewed the code, but tried it out and confirmed that it cleaned up the temp dir in an OpenStudio-HPXML run. Thanks @jmarrec!

**Comment by tijcolem** [comment]
@jmarrec This looks great. Thank for fixing so quickly for @shorowit. We can get this into 3.5.1 as well.

**Pull request merged**